### DISCUSSION TITLE
React Native OAuth using google provider

### DISCUSSION THREAD

**Discussion by SkillAcademia** [pocketbase/pocketbase#2484] [Q&A] [answered]
I am trying to integrate Google Login to my react native app. I have the following code for the same. I have also added the EventSource to my global context using react-native-sse which seems to solve the problem of undefined 'EventSource'.

```
  const handleOAuthLogin = async () => {
    setLoading(true);

    try {
      const authData = await pb
        .collection("users")
        .authWithOAuth2({ provider: "google" });

      console.log(authData);
    } catch (error) {
      console.log(error);
      console.log(error?.originalError);
    }

    setLoading(false);
  };
```

But now I have getting another ClientError as following from the catch block of my code. An also the google login screen doesn't pop up in my app.

```
[ClientResponseError 0: Something went wrong while processing your request.]
null
```

**Comment by ganigeorgiev** [comment]
I'm not sure what could be the cause of the error. If the above is called multiple times it could be caused by the [autocancellation](https://github.com/pocketbase/js-sdk#auto-cancellation). You can try to temp disable it globally with `pb.autoCancellation(false)` to see if that will resolve the issue.

> An also the google login screen doesn't pop up in my app.

The JS SDK by default uses `window.open` for launching a browser window and depending on which platform you target it may not be supported. You can provide your own browser launcher using the `urlCallback` prop:

```js
import { Linking } from "react-native";

await pb.collection("users").authWithOAuth2({
    provider: "google",
    urlCallback: (url) => {
        Linking.openURL(url).catch((err) => {
            console.log("failed to open url", err)
        })
    },
});
```

**Reply by SkillAcademia** [reply]
Thank you very much.

**Reply by SkillAcademia** [reply]
Hey, I tried implementing the same but no code inside the `urlCallback` is getting called.

```
const authData = await pb.collection("users").authWithOAuth2({
        provider: "google",
        urlCallback: (url) => {
          console.log("Hello");
          WebBrowser.openAuthSessionAsync(url).catch((err) => {
            console.log("failed to open url", err);
          });
        },
      });
```

**Reply by ganigeorgiev** [reply]
Could you create a minimal reproducible repo, for example via https://snack.expo.dev/?

**Reply by SkillAcademia** [reply]
So, I tried my code in snack and its working flawlessly. Let me see if I am getting this issue in my other android devices or not and also create a minimal reproducible repo having expo prebuild, since my original app is using expo prebuild to work with native modules. 

Here is the working snack: https://snack.expo.dev/@gaganbiswas/cc6c3f

**Reply by gaganbiswas** [reply]
@ganigeorgiev I figured out that the problem is arising only when we are prebuilding our expo app with ```npx expo prebuild```. Here's an example repo for the error: https://github.com/gaganbiswas/pocketbase-rn-oauth. Can you look into it and see what's the problem. The steps are included in readme of the repo.

**Reply by ganigeorgiev** [reply]
I don't have a modern Android device or proper emulator setup at the moment to test it. Is there a particular error that is thrown?

Note that for the PocketBase OAuth2 to work it doesn't really matter how the browser url was opened as long as the SSE connection was previously established in your app. To rule out a JS SDK issue you can try to leave only a print statement inside the `urlCallback` to verify that the function is called. If it is not called, then wrap the `handleLogin` content with `try/catch` statement to check if there is any error thrown.

**Comment by gaganbiswas** [comment]
Hey I tried what you asked and below is the result of the same. I added the comments which are the results I am getting from the console logs. I am not getting any information that might help me debug the error. In prebuild version the `urlCallback` function is not getting called.

```
import React from "react";
import { Button, StyleSheet, Text, View } from "react-native";
// import { openAuthSessionAsync } from "expo-web-browser";
import EventSource from "react-native-sse";
global.EventSource = EventSource;

import PocketBase from "pocketbase";
const pb = new PocketBase("https://skillacademia.fly.dev");

export default function Page() {
  const handleLogin = async () => {
    //Getting called
    console.log("Inside handle login");

    await pb
      .collection("users")
      .authWithOAuth2({
        provider: "google",
        urlCallback: async (url) => {
          // Not getting called. No error is thrown.
          console.log("Callback Url");
        },
      })
      .catch((err) => {
        // Pocketbase auth error [ClientResponseError 0: Something went wrong while processing your request.]
        console.log("Pocketbase auth error", err);
        // Pocketbase auth error null
        console.log("Pocketbase auth error", err?.originalError);
      });
  };

  return (
    <View style={styles.container}>
      <View style={styles.main}>
        <Text style={styles.title}>Pocketbase RN OAuth</Text>
        <Button
          title="Continue with Google"
          onPress={() => {
            try {
              handleLogin();
            } catch (error) {
              //Not getting called
              console.log("Error from button", error.message);
            }
          }}
        />
      </View>
    </View>
  );
}
```

**Reply by ganigeorgiev** [reply]
The `onPress` `try/catch` will work only if you await the async call. But I doubt that you'll catch anything as you already have `catch` inside the `handleLogin` method.

Is only the `authWithOauth2` call failing? Are you able to send regular collection requests?

Searching for possible issues in the code, I guess it is possible that maybe the `new URL(...)` constructor is causing it to fail (_at least based on https://github.com/facebook/react-native/issues/16434#issuecomment-557956079_). You can try to install a polyfill as one of the comments suggested to see if that will fix the issue - https://github.com/charpeni/react-native-url-polyfill.

Later this week I'll try to fix the errors wrapping and investigate why the `err.originalError` is `null`.

**Reply by ganigeorgiev** [reply]
I've replaced the `URL` with manual parsing to avoid the need of a polyfill in [JS SDK v0.15.2](https://github.com/pocketbase/js-sdk/releases/tag/v0.15.2).

The errors wrapping should also be fixed, but note that `null` could still be returned in some rare cases if the error is not serializable (or when there is no `originalError`).

@gaganbiswas, @SkillAcademia Please test again with the new version.

**Reply by gaganbiswas** [reply]
I tried with the pocketbase js-sdk v0.15.2 and now I am getting a new message in the original error: `[Error: EventSource connect took too long.]`. The code is unchanged and the code inside urlCallback is yet not getting called. I have previously tried with react-native-url-polyfill also and it didn't work.

Also, as a side note the code still works flawlessly in expo go app. It just doesn't work when we are prebuilding the app with `expo prebuild` to use native features and the react-native-sse is installed and binded to global.EventSource which works in fetching realtime data of the user.

**Reply by ganigeorgiev** [reply]
The _"EventSource connect took too long."_ is thrown when the SDK can't establish a SSE connection with the server in less than 15secs.

I'm still not sure what is the cause of the issue in your case, but I suspect that it could be related to the `react-native-sse` polyfill. There is a similar issue in https://github.com/binaryminds/react-native-sse/issues/11 but the author of the issue closed it for some reason. I'm not sure what to suggest other than to try with a different `EventSource` polyfill.

_Note that there are plans to remove the `EventSource` dependency from the JS SDK and use streams with `fetch`, but the Go JS app hooks are currently with higher priority._

**Reply by ganigeorgiev** [reply, marked as answer]
You may also want to check https://github.com/binaryminds/react-native-sse/issues/3 and the suggestion in https://stackoverflow.com/questions/69235694/react-native-cant-connect-to-sse-in-android/69235695#answer-69235695

_It seems that SSE in React Native on Android is a common issue - https://github.com/facebook/react-native/issues/28835._

**Reply by gaganbiswas** [reply]
Thank you! It now works perfectly. The issue was from the NetworkFlipper in debug mode. Commenting out the same as described in [https://stackoverflow.com/questions/69235694/react-native-cant-connect-to-sse-in-android/69235695#answer-69235695](https://stackoverflow.com/questions/69235694/react-native-cant-connect-to-sse-in-android/69235695#answer-69235695) helped in resolving the issue. 

It seems the problem is only in debug mode and is caused due to NetworkFlipper and happens only in Android.

**Comment by brunowilliang** [comment]
Hi guys, I'm trying to login through github and I'm not succeeding, this is the error

```shel
 WARN  Possible Unhandled Promise Rejection (id: 0):
ClientResponseError 0: Something went wrong while processing your request.
ClientResponseError@http://192.168.100.58:19000/index.tsx.bundle?platform=ios&dev=true&hot=false:649011:27
@http://192.168.100.58:19000/index.tsx.bundle?platform=ios&dev=true&hot=false:649711:55
step@http://192.168.100.58:19000/index.tsx.bundle?platform=ios&dev=true&hot=false:648988:23
@http://192.168.100.58:19000/index.tsx.bundle?platform=ios&dev=true&hot=false:648999:10
fulfilled@http://192.168.100.58:19000/index.tsx.bundle?platform=ios&dev=true&hot=false:648903:22
tryCallOne@http://192.168.100.58:19000/index.tsx.bundle?platform=ios&dev=true&hot=false:28689:16
@http://192.168.100.58:19000/index.tsx.bundle?platform=ios&dev=true&hot=false:28770:27
@http://192.168.100.58:19000/index.tsx.bundle?platform=ios&dev=true&hot=false:29665:26
_callTimer@http://192.168.100.58:19000/index.tsx.bundle?platform=ios&dev=true&hot=false:29584:17
_callReactNativeMicrotasksPass@http://192.168.100.58:19000/index.tsx.bundle?platform=ios&dev=true&hot=false:29614:17
callReactNativeMicrotasks@http://192.168.100.58:19000/index.tsx.bundle?platform=ios&dev=true&hot=false:29777:44
__callReactNativeMicrotasks@http://192.168.100.58:19000/index.tsx.bundle?platform=ios&dev=true&hot=false:2696:46
@http://192.168.100.58:19000/index.tsx.bundle?platform=ios&dev=true&hot=false:2508:45
__guard@http://192.168.100.58:19000/index.tsx.bundle?platform=ios&dev=true&hot=false:2680:15
flushedQueue@http://192.168.100.58:19000/index.tsx.bundle?platform=ios&dev=true&hot=false:2507:21
flushedQueue@[native code]
invokeCallbackAndReturnFlushedQueue@[native code]
```

My github settings
<img width="555" alt="Screenshot 2023-07-05 at 03 46 14" src="https://github.com/pocketbase/pocketbase/assets/41432581/b24b64bd-b7b3-4b4b-a25e-c6da9e0bc05d">

My Code:
```js
      githubLogin: async () => {
        await api.collection(Collections.Users).authWithOAuth2({
          provider: 'github',
          urlCallback: async (url) => {
            WebBrowser.openAuthSessionAsync(url).catch((err) => {
              console.log('failed to open url', err)
            })
          },
        })
      },
```

can someone help me?

**Reply by ganigeorgiev** [reply]
It is hard to tell what could be the reason for the error in your case.
You can try to wrap with try/catch the `await` call and `console.log(err.originalError)` to inspect the exception for more details.

**Reply by brunowilliang** [reply]
@ganigeorgiev it worked now, all that was left was to install "react-native-url-polyfill" hahaha

another question, is there a way to check if the returned token is valid? because I try to get:
```js
   const token2 = api.authStore.token
   const isValid2 = api.authStore.isValid
```
it works the first time but when I reload it it doesn't come back empty

**Reply by ganigeorgiev** [reply]
I'm not sure I follow. If you mean that it is not persisted, then Yes, `pb.authStore` will try to use `localStorage` if available, otherwise will fallback to only runtime store. For a react native custom AuthStore example you can check https://github.com/pocketbase/pocketbase/discussions/1856#discussioncomment-4986746.

**Reply by brunowilliang** [reply]
@ganigeorgiev I'm doing just that.
the problem is that when i revoke access in github:
<img width="858" alt="Screenshot 2023-07-05 at 15 36 37" src="https://github.com/pocketbase/pocketbase/assets/41432581/6b7cb169-2300-4feb-9b1c-c02f59221735">

even so it gives in my console:
``` shell
LOG {
   "token": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJjb2xsZWN0aW9uSWQiOiJfcGJfdXNlcnNfYXV0aF8iLCJleHAiOjE2ODk3NzczMDgsImlkIjoiY2Q1Z3hxcHE2NDI0aGc4 IiwidHlwZSI6ImF1dGhSZWNvcmQifQ.aFY_g-pw-fIw07sp2GaoyITGGnegIloQCGs_ewcvyUo",
   "isValid": true
}
```

**Reply by ganigeorgiev** [reply]
The GitHub and PocketBase auth tokens are 2 different things and they operate independently.

Revoking or even deleting the GitHub OAuth2 app will not invalidate previously issued PocketBase auth tokens because we don't have any information about this change.

**Reply by brunowilliang** [reply]
omg, sorry, I understand

so he will validate based on that?
<img width="844" alt="Screenshot 2023-07-05 at 15 52 03" src="https://github.com/pocketbase/pocketbase/assets/41432581/a8bb8bab-ab34-4922-a51e-4d46280139f3">

**Reply by ganigeorgiev** [reply]
Yes, the PocketBase auth tokens expiration is configured from this page.

**Reply by brunowilliang** [reply]
wonderful, it worked perfectly

tkss @ganigeorgiev

**Reply by brunowilliang** [reply]
@ganigeorgiev just one more question, is there a way to save the avatar at login?

**Reply by ganigeorgiev** [reply]
_@Brunowilliang Please if you have other questions open a new discussion as this is already hard to follow and will make it difficult to search and find if someone else is having the same issue._

In general, the only OAuth2 fields we persist with the user creation are the `email` (if available) and the `username` (if available and complies with the PocketBase username format). For the avatar, you'll have to download the image on your own (if available it will be returned under the `meta.avatarUrl` key) and send a new regular `update()` request for your user.

**Comment by kingdevnl** [comment]
Are there any working examples for a react-native app on how to setup OAuth?

**Comment by manish181192** [comment]
This is a working example of oauth2 login with google. But my browser is not closing automatically. @ganigeorgiev @gaganbiswas Can you help me out on how to close the web browser?

@kingdevnl 

```
import { StatusBar } from 'expo-status-bar';
import { Button, StyleSheet, Text, TextInput, TouchableOpacity, View } from 'react-native';
import authStore from './AuthStore';
import PocketBase from "pocketbase";
import { useState } from 'react';
import { Linking } from "react-native";
import 'react-native-url-polyfill/auto';
import * as WebBrowser from "expo-web-browser";
import EventSource from "react-native-sse";

global.EventSource = EventSource;

export default function App() {
  const isValid = authStore((state) => (state.isValid)); 
  const setIsvalid = authStore((state) => (state.setIsValid));
  const setIsInvalid = authStore((state) => state.setIsInvalid);

  const pb = new PocketBase("http://localhost:8090");

  const [email, setEmail] = useState("");
  const [password, setPassword] = useState("");
  const [emailError, setEmailError] = useState("");
  const [passwordError, setPasswordError] = useState("");
  const [signinError, setSigninError] = useState("");
  
  
  async function handleLogout() {
    setIsInvalid();
    pb.authStore.clear();
    console.log("loggedout" , isValid);
  }

  async function handleLogin() {
    console.log(email);
    console.log(password);
    try {
      const authData = await pb.collection('users').authWithPassword(
        email,
        password,
      )
      console.log(authData);
      setIsvalid();
    } catch (error) {
      if (error.response && error.response.data) {
        if (error.response.data != null && error.response.data.identity != null) {

          setEmailError(error.response.data.identity.message);
        } else {
          setEmailError("");
        }
        if (error.response.data != null && error.response.data.password != null) {
          setPasswordError(error.response.data.password.message);
        } else {
          setPasswordError("");
        }
      }
      if (error.response != null && error.response.message != null) {
        setSigninError(error.response.message);
        console.log("signing error");
        console.log(signinError);
      }

      console.log("error", error);
    }

  }

  async function handleGoogleLogin() {
    // const url = "http://localhost:8090/api/oauth2-redirect";
    try {
      pb.autoCancellation(false)
      // const authData = await pb.collection('users').authWithOAuth2({ provider: 'google' });
      await pb.collection("users").authWithOAuth2({
        provider: "google",
        urlCallback: (url) => {
          Linking.openURL(url).catch((err) => {
            console.log("failed to open url", err)
          })
        },
      });
      setIsvalid();
    } catch (e) {
      console.log(e);
    }

  }

  return (
    <View style={styles.container}>
    {
      !isValid?
       ( 
       <>
        <Text>Login</Text>
              <TextInput style={styles.email} autoCapitalize='none' placeholder='enter email...' textContentType='emailAddress' onChangeText={(text) => setEmail(text)} />
              <TextInput style={styles.password} autoCapitalize='none' placeholder='enter password...' textContentType='password' onChangeText={(text) => setPassword(text)} secureTextEntry={true} />
        <TouchableOpacity style={styles.button} onPress={handleLogin}>
          <Text> Login </Text>
        </TouchableOpacity>
        <TouchableOpacity style={styles.button} onPress={handleGoogleLogin}>
          <Text> Sign-in with Google </Text>
        </TouchableOpacity>
        <StatusBar style="auto" />
       </>
       )
       : 
       (
        <>
        <Text> Home </Text>
        <TouchableOpacity style={styles.button} onPress={handleLogout}>
          <Text> Logout </Text>
        </TouchableOpacity>
        </>
       ) 
    }
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#fff',
    alignItems: 'center',
    justifyContent: 'center',
  },
  email:{},
  password:{
  },
  button: {
    padding: 10,
    borderRadius: 4,
    borderWidth: 1,
    borderColor: 'green',
    backgroundColor: 'lightgreen',
  }

});
```

**Reply by ganigeorgiev** [reply]
In general the browser window will be automatically closed if it is allowed but that's not always the case (_see also [window.close docs](https://developer.mozilla.org/en-US/docs/Web/API/Window/close)_).
Or in other words, it may not be possible and it will depend on the platform, browser and how it was opened, etc.
This is why the users should see a message like: _"You can go back to the app if **this window is not automatically closed**"_.
